### DISCUSSION TITLE
JSON RPC parse issue

### DISCUSSION THREAD

**Discussion by vasylmykytiuk** [AsamK/signal-cli#1144] [Q&A] [answered]
Hey,
Thanks for the great tool for using Signal protocol.

Currently, I'm trying to integrate signal-cli with my dotnet application using https://github.com/microsoft/vs-streamjsonrpc library.
```
        var cliProcessInfo = new ProcessStartInfo("signal-cli.bat")
        {
                Arguments = "-u +<my_number> jsonRpc",
                RedirectStandardInput = true,
                RedirectStandardOutput = true,
                UseShellExecute = false,
                CreateNoWindow = true
        };        
        Process cliProcess = Process.Start(cliProcessInfo);

        var formatter = new JsonMessageFormatter(Encoding.UTF8);

        var messageHandler = new NewLineDelimitedMessageHandler(cliProcess.StandardInput.BaseStream, cliProcess.StandardOutput.BaseStream, formatter)
        {
            NewLine = NewLineDelimitedMessageHandler.NewLineStyle.Lf
        };

        var jsonRpc = new StreamJsonRpc.JsonRpc(messageHandler);
        jsonRpc.StartListening();

        try
        {
            await jsonRpc.InvokeAsync("listGroups");
        }
        catch (Exception e)
        {
            Console.WriteLine(e);
        }
```

It fails with the following error:
```
{
  "jsonrpc": "2.0",
  "error": {
    "code": -32700,
    "message": "Unrecognized token '�': was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n at [Source: (String)\"﻿{\"jsonrpc\":\"2.0\",\"id\":2,\"method\":\"listGroups\",\"params\":[]}\"; line: 1, column: 2]",
    "data": null
  },
  "id": null
}
```

This invoke `await jsonRpc.InvokeAsync("listGroups");` writes to the input stream the following JSON:
`{"jsonrpc": "2.0","id": 2,"method": "listGroups","params": []}` - please not the NewLineDelimitedMessageHandler writes '\n' character at the end to recognize the message 'boarders' - is it correct format for signal-cli?

Many thanks for any help.

**Comment by AsamK** [comment, marked as answer]
Your line has a Byte Order Mark (ef bb  bf) in the beginning. signal-cli doesn't expect that, as the input is always UTF-8.
You need to check your library how to remove the BOM.

**Reply by vasylmykytiuk** [reply]
Thanks for your answer.
Imagine that I will remove the BOM and write to steam only the bytes of JSON RPC request message.
What is the format of JSON RPC response message? How should I know that the response is completed?

**Reply by AsamK** [reply]
The response message is a single JSON object, terminated by a new line delimiter ('\n').
The ordering of the responses is not guaranteed, so you should check for the response with the `id` field matching your request.

**Reply by vasylmykytiuk** [reply]
Ok, for the ones who facing the same issue I resolved it by this:
```
var formatter = new JsonMessageFormatter(new UTF8Encoding(false));

var messageHandler = new NewLineDelimitedMessageHandler(
      cliProcess.StandardInput.BaseStream,
      cliProcess.StandardOutput.BaseStream,
      formatter)
{
      NewLine = NewLineDelimitedMessageHandler.NewLineStyle.Lf
};

var jsonRpc = new JsonRpc(messageHandler);
```

Then you can query for groups:
`JsonGroup[] result = await JsonRpc.InvokeWithParameterObjectAsync<JsonGroup[]>("listGroups");`
